deployer/src/lib.rs

```diff
@@ -1,4 +1,604 @@
+                state_variant: creating.to_string(),
+                state: creating,
+            };
+            self.dal
+                .insert_service_if_absent(service)
+                .await
+                .map_err(Error::Dal)?;
+
+            // Insert the new deployment.
+            let deployment = Deployment {
+                id: runnable_deployment.deployment_id,
+                service_id: runnable_deployment.service_id,
+                last_update: Utc::now(),
+                is_next: runnable_deployment.is_next,
+                git_branch: git_metadata.git_branch,
+                git_commit_hash: git_metadata.git_commit_hash,
+                git_commit_message: git_metadata.git_commit_message,
+                git_dirty: git_metadata.git_dirty,
+            };
+            self.dal.insert_deployment(deployment).await?;
+        }
+
+        // We want to refresh the service.
+        let service_id = runnable_deployment.service_id;
+        let cs = ContainerSettings::builder()
+            .image(image_name)
+            .provisioner_host(self.config.provisioner_uri.to_string())
+            .auth_uri(self.config.auth_uri.to_string())
+            .network_name(self.config.network_name.to_string())
+            .runnable_deployment(runnable_deployment)
+            .runtime_start_channel(self.runtime_start_channel.clone())
+            .prefix(self.config.prefix.to_string())
+            .build()
+            .await;
+
+        TaskBuilder::new(self.dal.clone())
+            .task_router(self.task_router.clone())
+            .service_id(service_id)
+            .service_docker_context(ServiceDockerContext::new_with_container_settings(
+                self.docker.clone(),
+                cs,
+                self.runtime_manager.clone(),
+            ))
+            .and_then(task::refresh())
+            .and_then(task::run_until_done())
+            .send(&self.deployment_state_machine_channel)
+            .await
+            .expect("to get a handle of the created task")
+            .await;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl<D: Dal + Sync + 'static> Deployer for DeployerService<D> {
+    #[instrument(skip(self, request))]
+    async fn deploy(
+        &self,
+        request: tonic::Request<DeployRequest>,
+    ) -> TonicResult<tonic::Response<DeployResponse>, tonic::Status> {
+        // Authorize the request.
+        request.verify(Scope::DeploymentWrite)?;
+
+        let claim = request.extensions().get::<Claim>().cloned();
+        let request = request.into_inner();
+        let req_deployment = match request.deployment {
+            Some(inner) => inner,
+            None => {
+                return Ok(Response::new(DeployResponse {
+                    success: false,
+                    message: Some("invalid argument: missing deployment information".to_string()),
+                    deployment_id: None,
+                }))
+            }
+        };
```

Comment:
This and the one below could use let else, e.g. `let Ok(req_deployment) = request.deployment else { return ...}`